networkit/cpp/generators/ConfigurationModel.cpp

```diff
@@ -0,0 +1,85 @@
+
+#include <networkit/generators/ConfigurationModel.hpp>
+
+#include <networkit/auxiliary/Random.hpp>
+#include <networkit/generators/HavelHakimiGenerator.hpp>
+#include <networkit/graph/Graph.hpp>
+
+#include <iostream>
+
+namespace NetworKit {
+
+ConfigurationModel::ConfigurationModel(const std::vector<count> &sequence)
+    : StaticDegreeSequenceGenerator(sequence) {
+
+    // Throw an error if the sequence isn't realizable:
+    std::sort(seq.begin(), seq.end(), std::greater<count>());
+    HavelHakimiGenerator(seq).generate();
+}
+
+Graph ConfigurationModel::generate() {
+    do {
+        Graph result{seq.size()};
+
+        std::vector<index> urn{};
+        std::vector<count> remaining(seq.size());
+        std::vector<bool> alreadyTaken(seq.size(), false);
+        index remainingNodes = seq.size();
+
+        // Fill the urn:
+        for (index i = 0; i < seq.size(); ++i) {
+            for (count j = 0; j < seq[i]; ++j) {
+                urn.push_back(i);
+            }
+            remaining[i] = seq[i];
+            if (seq[i] == 0) {
+                --remainingNodes;
+            }
+        }
+
+        count remainingSize = urn.size();
+        index curIndex = 0;
+
+        auto sample = [&remainingSize, &urn, &alreadyTaken, &curIndex]() -> index {
+            index result;
+            do {
+                --remainingSize;
+                index i;
+                do {
+                    i = Aux::Random::integer(remainingSize);
+                    result = urn[i];
+                } while (alreadyTaken[result]);
```

Comment:
This becomes very inefficient if many elements are already taken. Why don't we keep a `std::vector<index>` of available indices, sample one element from it using `std::sample`, and remove the sampled element from the sequence?

Reply:
As far as i could test, this way is more efficient, here's why: We are sampling from the urn (which represents the degree seq in the following way: If the largest degree is 36, there are 36 `0` entries, second largest degree is 22, there are 22 `1` entries and so on ..) 

When sampling from the urn, the probability of being sampled is proportional to the degree of the nodes. Therefore, when we want to sample in a similar fashion, we need a practical copy of the urn, which only contains "available indices". Updating this rather large vector for every sample takes much longer than re-sampling (according to local testing). 

If you have any other ideas for a more efficient implementation, feel free to elaborate.